### ISSUE TITLE
Use mutable objects in classes instead of indexing _args

### ISSUE THREAD

**Issue by EmilyBourne** [pyccel/pyccel#660]
Most classes use the `_args` variable (or `.args` property) provided by the sympy inheritance to store their arguments. This means that the arguments are immutable and several of our properties are hard to maintain, e.g:
```
@property
def thing(self):
    return self._args[10]
```

E.g.
```
class While(Basic):
    def __new__(cls, test, body, local_vars=[]):
        test = sympify(test, locals=local_sympify)

        if PyccelAstNode.stage == 'semantic':
            if test.dtype is not NativeBool():
                test = PythonBool(test)

        if iterable(body):
            body = CodeBlock((sympify(i, locals=local_sympify) for i in body))
        elif not isinstance(body,CodeBlock):
            raise TypeError('body must be an iterable or a CodeBlock')
        return Basic.__new__(cls, test, body, local_vars)

    @property
    def test(self):
        return self._args[0]

    @property
    def body(self):
        return self._args[1]

    @property
    def local_vars(self):
        return self._args[2]
```
Should instead be:
```
class While(Basic):
    def __init__(self, test, body, local_vars=[]):
        test = sympify(test, locals=local_sympify)

        if PyccelAstNode.stage == 'semantic':
            if test.dtype is not NativeBool():
                test = PythonBool(test)

        if iterable(body):
            body = CodeBlock((sympify(i, locals=local_sympify) for i in body))
        elif not isinstance(body,CodeBlock):
            raise TypeError('body must be an iterable or a CodeBlock')
            
        self._test = test
        self._body = body
        self._local_vars = local_vars
        Basic.__init__(self)

    @property
    def test(self):
        return self._test

    @property
    def body(self):
        return self._body

    @property
    def local_vars(self):
        return self._local_vars
```

This needs fixing for all files in the folder pyccel/ast:
- [x] basic
- [x] bind_c
- [x] bitwise_operators.py
- [ ] builtins.py
- [x] core.py
- [x] cwrapper.py
- [ ] datatypes.py
- [x] functionalexpr.py
- [x] headers.py
- [x] internals.py
- [x] itertoolsext.py
- [ ] literals.py
- [ ] macros.py
- [x] mathext.py
- [ ] numpyext.py
- [x] operators.py
- [x] sympy_helper.py
- [x] utilities.py
- [ ] variable.py

**Comment by EmilyBourne** [comment]
I am working on fixing this in `ast/core.py` in the branch ebourne_mutable_core. I am not assigning the issue to myself as there are many other files for which the same thing can be done simultaneously

**Comment by EmilyBourne** [comment]
Some of this is fixed in PR #718 but this should be completed before to make the review process manageable
